src/domain_bridge/wait_for_qos_handler.hpp

```diff
@@ -0,0 +1,276 @@
+   *   for the provided topic name.
+   *   If QoS settings are already available, then the callback is called immediately.
+   */
+  void register_on_publisher_qos_ready_callback(
+    std::string topic,
+    rclcpp::Node::SharedPtr node,
+    std::function<void(const QosMatchInfo)> callback)
+  {
+    // If the QoS is already avaiable, trigger the callback immediately
+    auto opt_qos = get_topic_qos(topic, node);
+    if (opt_qos) {
+      const QosMatchInfo & qos = opt_qos.value();
+      callback(qos);
+      return;
+    }
+
+    {
+      std::lock_guard<std::mutex> lock(waiting_map_mutex_);
+      waiting_map_[node].push_back({topic, callback});
+
+      // If we already have a thread for this node, then notify that there is a new callback
+      auto existing_thread = waiting_threads_.find(node);
+      if (existing_thread != waiting_threads_.end()) {
+        existing_thread->second.second->notify_all();
+        return;
+      }
+    }
+
+    // If we made it this far, there doesn't exist a thread for waiting so we'll create one
+    // First, create a condition variable to notify the waiting thread when to wake up
+    auto cv = std::make_shared<std::condition_variable>();
+
+    // Create a thread that waits for a publisher to become available
+    auto invoke_callback_when_qos_ready = [this, node, cv]()
+      {
+        auto event = node->get_graph_event();
+        while (!this->shutting_down_.load()) {
+          // Wait for graph event
+          // Note, in case new publishers don't trigger a graph event we add a
+          // timeout so that we can still poll periodically for new publishers
+          node->wait_for_graph_change(event, std::chrono::seconds(1));
+          event->check_and_clear();
+          // Check if QoS is ready for any of the topics
+          {
+            std::unique_lock<std::mutex> lock(waiting_map_mutex_);
+            auto & topic_callback_vec = waiting_map_.at(node);
+            for (auto it = topic_callback_vec.begin(); it != topic_callback_vec.end(); ++it) {
+              const std::string & topic = it->first;
+              const auto callback = it->second;
+              auto opt_qos = this->get_topic_qos(topic, node);
+              if (opt_qos) {
+                const QosMatchInfo & qos = opt_qos.value();
+                callback(qos);
+                topic_callback_vec.erase(it--);
+              }
+            }
+            // If there are no more publishers to wait on,
+            // remove the entry from the map and terminate the thread
+            if (topic_callback_vec.size() == 0u) {
+              waiting_map_.erase(waiting_map_.find(node));
+            }
+
+            // It's only worth continuing if we have callbacks to handle,
+            // or we're shutting down
+            cv->wait(
+              lock,
+              [this, &topic_callback_vec]
+              {return topic_callback_vec.size() > 0u || this->shutting_down_.load();});
```

Comment:
don't combine condition variables and atomics, this code is equivalent to:

```cpp
while (!(topic_callback_vec.size() > 0u || this->shutting_down_.load())) {
  // boolean flipped here and notification sent here, deadlock
  cv->wait(lock);
}
```

[this article](https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables) explains the problem well.

tl;dr: replace the atomic with a normal `bool` and lock the `waiting_map_mutex_` before flipping it (no need to hold the mutex while calling `notify_all()` though).

Reply:
actually because of the logic used here, it seems that `topic_callback_vec.size() > 0u` is always true.
Maybe just deleting the condition variable code (?)

Reply:
I think without the condition variable, the thread essentially runs a busy-loop when `topic_callback_vec.size() == 0u`. There is no work to do, but it will continue looping and waiting for graph events.

Reply:
Thanks for catching the deadlock bug! If I remove the atomic from the predicate, I get other shutdown issues; Looking into it.

Reply:
> I think without the condition variable, the thread essentially runs a busy-loop when topic_callback_vec.size() == 0u. There is no work to do, but it will continue looping and waiting for graph events.

True, I didn't check that correctly.

Reply:
> Thanks for catching the deadlock bug! If I remove the atomic from the predicate, I get other shutdown issues; Looking into it.

Just to be more clear, my recommendation was to replace this line

https://github.com/ros2/domain_bridge/blob/4ae1095a59b290fdd72ec68451ef66d76ab8afdc/src/domain_bridge/wait_for_qos_handler.hpp#L67

with

```cpp
{
  std::lock_guard<std::mutex> lock(waiting_map_mutex_);
  shutting_down_ = true; 
}
```

and `shutting_down_` can be made a `bool`.

Reply:
Now using a regular bool for `shutting_down_` and fixed some cleanup logic in the event of a SIGINT: eac644d